src/libraries/Microsoft.Extensions.Logging.Console/src/AnsiParser.cs

```diff
@@ -0,0 +1,207 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+
+namespace Microsoft.Extensions.Logging.Console
+{
+    internal class AnsiParser
+    {
+        private readonly Action<string, int, int, ConsoleColor?, ConsoleColor?> _onParseWrite;
+        public AnsiParser(Action<string, int, int, ConsoleColor?, ConsoleColor?> onParseWrite)
+        {
+            if (onParseWrite == null)
+            {
+                throw new ArgumentNullException(nameof(onParseWrite));
+            }
+            _onParseWrite = onParseWrite;
+        }
+
+        /// <summary>
+        /// Parses a subset of display attributes
+        /// Set Display Attributes
+        /// Set Attribute Mode [{attr1};...;{attrn}m
+        /// Sets multiple display attribute settings. The following lists standard attributes that are getting parsed:
+        /// 1 Bright
+        /// Foreground Colours
+        /// 30 Black
+        /// 31 Red
+        /// 32 Green
+        /// 33 Yellow
+        /// 34 Blue
+        /// 35 Magenta
+        /// 36 Cyan
+        /// 37 White
+        /// Background Colours
+        /// 40 Black
+        /// 41 Red
+        /// 42 Green
+        /// 43 Yellow
+        /// 44 Blue
+        /// 45 Magenta
+        /// 46 Cyan
+        /// 47 White
+        /// </summary>
+        public void Parse(string message)
+        {
+            int startIndex = -1;
+            int length = 0;
+            ConsoleColor? foreground = null;
+            ConsoleColor? background = null;
+            var span = message.AsSpan();
+            const char EscapeChar = '\x1B';
+            ConsoleColor? color = null;
+            bool isBright = false;
+            for (int i = 0; i < span.Length; i++)
+            {
+                if (span[i] == EscapeChar && span.Length >= i + 4 && span[i + 1] == '[')
+                {
+                    if (span[i + 3] == 'm')
+                    {
+    #if (NETSTANDARD2_0 || NETFRAMEWORK)
+                        if (ushort.TryParse(span.Slice(i + 2, length: 1).ToString(), out ushort escapeCode))
```

Comment:
nit: we may enhance this later by manually parsing the number as it should be straights forward and faster too.

Reply:
We could, but it would only be for .NET Framework, and older .NET Core versions (which people won't use 5.0 Microsoft.Extensions on older .NET Core versions by default). So I don't think it is critical to optimize this case.

---
In reply to: [454691144](https://github.com/dotnet/runtime/pull/38616#discussion_r454691144) [](ancestors = 454691144)